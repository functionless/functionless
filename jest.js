const swcJest = require("@swc/jest");
const { config } = require("./swc-config");

function createTransformer() {
  return swcJest.createTransformer(config);
}

module.exports = { createTransformer };
