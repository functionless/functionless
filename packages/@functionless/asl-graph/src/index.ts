export * from "./asl-graph";
export * from "./asl-integration";
export * from "./asl";
export * from "./eval-expr-context";
export * from "./guards";
export * from "./states";
export * from "./step-function-error";
