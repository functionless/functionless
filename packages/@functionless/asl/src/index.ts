export { ASLGraph } from "./asl-graph";
export { ASLIntegration, isASLIntegration } from "./asl-integration";
export { ASL } from "./asl";
export * from "./eval-expr-context";
export * from "./guards";
export {
  Branch,
  Catch,
  Choice,
  CommonFields,
  CommonTaskFields,
  Condition,
  Fail,
  MapTask,
  ParallelTask,
  Parameters,
  Pass,
  Retry,
  State,
  StateMachine,
  States,
  Succeed,
  Task,
  TerminalState,
  Wait,
  isChoiceState,
  isFailState,
  isMapTaskState,
  isParallelTaskState,
  isPassState,
  isState,
  isSucceedState,
  isTaskState,
  isWaitState,
} from "./states";
export { StepFunctionCause, StepFunctionError } from "./step-function-error";
