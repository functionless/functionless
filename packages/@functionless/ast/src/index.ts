export * from "./assert";
export * from "./constants";
export * from "./declaration";
export * from "./ensure";
export * from "./error";
export * from "./expression";
export * from "./find";
export * from "./guards";
export * from "./literal";
export * from "./node-clone";
export * from "./node-ctor";
export * from "./node-kind";
export * from "./node";
export * from "./reflect";
export * from "./resolve-references";
export * from "./s-expression";
export * from "./span";
export * from "./statement";
export * from "./util";
export * from "./visit";
