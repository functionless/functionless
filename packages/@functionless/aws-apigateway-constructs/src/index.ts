export { APIGatewayVTL } from "./api-vtl";
export {
  APIGatewayError,
  APIGatewayRequestOverride,
  ApiBody,
  ApiGatewayContext,
  ApiGatewayInput,
  ApiGatewayResponseOverride,
  ApiMethod,
  ApiMethodKind,
  ApiParameter,
  ApiParameters,
  ApiRequest,
  AwsMethod,
  HttpMethod,
  LambdaMethod,
  LambdaMethodProps,
  MethodProps,
  MockMethod,
  isApiMethodKind,
} from "./api";
