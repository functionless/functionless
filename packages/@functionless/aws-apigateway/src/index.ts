export * from "./api-vtl";
export * from "./api-integration";
