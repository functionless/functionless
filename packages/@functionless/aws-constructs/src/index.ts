export * from "../../aws-appsync-constructs/src/api";
export * from "./appsync";
export * from "./async-synth";
export * from "./aws";
export * from "./event-bridge";
export * from "./event-source";
export * from "./iterable";
export * from "./queue";
export * from "./queue";
export * from "./secret";
export * from "@functionless/aws-stepfunctions-constructs/src/step-function";
export * from "./table";
export * from "./user-pool";
export * from "./util";

// to prevent the closure serializer from trying to import all of functionless.
export const deploymentOnlyModule = true;
