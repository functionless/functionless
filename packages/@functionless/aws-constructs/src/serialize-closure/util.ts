import path from "path";
import { FunctionlessNode } from "@functionless/ast";
import { SourceNode } from "source-map";

export function undefinedExpr() {
  return "undefined";
}

export function nullExpr() {
  return "null";
}

export function trueExpr() {
  return "true";
}

export function falseExpr() {
  return "false";
}

export function idExpr(name: string) {
  return name;
}

export function stringExpr(name: string) {
  // this seems dangerous - are we handling this right?
  return `"${name.replaceAll('"', '\\"').replaceAll("\n", "\\n")}"`;
}

export function numberExpr(num: number) {
  return num.toString(10);
}

export function bigIntExpr(num: bigint) {
  return `${num.toString(10)}n`;
}

export function regExpr(regex: RegExp) {
  return `/${regex.source}/${regex.flags}`;
}

export function objectExpr(obj: Record<string, SourceNodeOrString>) {
  return createSourceNodeWithoutSpan(
    "{",
    ...Object.entries(obj).flatMap(([name, val]) => [
      createSourceNodeWithoutSpan(name, " : ", val),
      ",",
    ]),
    "}"
  );
}

export type SourceNodeOrString = string | SourceNode;

export function createSourceNodeWithoutSpan<S extends SourceNodeOrString[]>(
  ...exprs: S
): S extends string ? string : SourceNodeOrString {
  if (exprs.every((expr) => typeof expr === "string")) {
    return exprs.join("");
  } else {
    return new SourceNode(null, null, "index.js", exprs) as any;
  }
}

export function createSourceNode(
  node: FunctionlessNode,
  chunks: string | SourceNodeOrString[]
) {
  const absoluteFileName = node.getFileName();

  return new SourceNode(
    node.span[0],
    node.span[1],
    path.relative(process.cwd(), absoluteFileName),
    chunks
  );
}

const propNameRegex = /^[_a-zA-Z][_a-zA-Z0-9]*$/g;

export function propAccessExpr<S extends SourceNodeOrString>(
  expr: S,
  name: string
): S extends string ? string : SourceNodeOrString {
  if (name.match(propNameRegex)) {
    return createSourceNodeWithoutSpan(expr, ".", name) as S extends string
      ? string
      : SourceNodeOrString;
  } else {
    return createSourceNodeWithoutSpan(
      expr,
      "[",
      stringExpr(name),
      "]"
    ) as S extends string ? string : SourceNodeOrString;
  }
}

export function assignExpr(
  left: SourceNodeOrString | SourceNode,
  right: SourceNodeOrString | SourceNode
) {
  return createSourceNodeWithoutSpan(left, " = ", right);
}

export function callExpr<
  E extends SourceNodeOrString,
  A extends SourceNodeOrString
>(expr: E, args: A[]): E | A extends string ? string : SourceNodeOrString {
  return createSourceNodeWithoutSpan(
    expr,
    "(",
    ...args.flatMap((arg, i) => (i < args.length - 1 ? [arg, ","] : [arg])),
    ")"
  ) as E | A extends string ? string : SourceNodeOrString;
}

export function newExpr(expr: string, args: string[]): string;
export function newExpr(expr: SourceNodeOrString, args: SourceNodeOrString[]) {
  return createSourceNodeWithoutSpan(
    "new ",
    expr,
    "(",
    ...args.flatMap((arg) => [arg, ","]),
    ")"
  );
}

export function exprStmt(expr: SourceNodeOrString | SourceNode) {
  return createSourceNodeWithoutSpan(expr, ";");
}

export function setPropertyStmt(
  on: SourceNodeOrString,
  key: string,
  value: SourceNodeOrString
) {
  return createSourceNodeWithoutSpan(setPropertyExpr(on, key, value), ";");
}

export function setPropertyExpr(
  on: SourceNodeOrString,
  key: string,
  value: SourceNodeOrString
) {
  return assignExpr(propAccessExpr(on, key), value);
}

export function definePropertyExpr(
  on: SourceNodeOrString,
  name: SourceNodeOrString,
  value: SourceNodeOrString
) {
  return callExpr(propAccessExpr(idExpr("Object"), "defineProperty"), [
    on,
    name,
    value,
  ]);
}

export function getOwnPropertyDescriptorExpr(
  obj: SourceNodeOrString,
  key: SourceNodeOrString
) {
  return callExpr(
    propAccessExpr(idExpr("Object"), "getOwnPropertyDescriptor"),
    [obj, key]
  );
}
