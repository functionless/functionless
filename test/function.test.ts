import { App, aws_lambda, Stack } from "aws-cdk-lib";
import "jest";
import { Function, AppsyncContext, reflect } from "../src";
import { VTL } from "../src/vtl";
import { appsyncTestCase } from "./util";

interface Item {
  id: string;
  name: number;
}

const app = new App({ autoSynth: false });
const stack = new Stack(app, "stack");

const lambda = new aws_lambda.Function(stack, "F", {
  code: aws_lambda.Code.fromInline(
    "exports.handler = function() { return null; }"
  ),
  handler: "index.handler",
  runtime: aws_lambda.Runtime.NODEJS_14_X,
});

const fn1 = Function.fromFunction<{ arg: string }, Item>(lambda);
const fn2 = Function.fromFunction<{ arg: string; optional?: string }, Item>(
  lambda
);
const fn3 = Function.fromFunction<undefined, Item>(lambda);
const fn4 = Function.fromFunction<{ arg: string }, void>(lambda);

test("call function", () =>
  appsyncTestCase(
    reflect((context: AppsyncContext<{ arg: string }>) => {
      return fn1(context.arguments);
    }),
    // pipeline's request mapping template
    "{}",
    // function's request mapping template
    `${VTL.CircuitBreaker}
#set($v1 = {\"version\": \"2018-05-29\", \"operation\": \"Invoke\", \"payload\": $context.arguments})
$util.toJson($v1)`,
    // function's response mapping template
    `#set( $context.stash.return__flag = true )
#set( $context.stash.return__val = $context.result )
{}`,
    // response mapping template
    `#if($context.stash.return__flag)
  #return($context.stash.return__val)
#end`
  ));

test("call function and conditional return", () =>
  appsyncTestCase(
    reflect((context: AppsyncContext<{ arg: string }>) => {
      const result = fn1(context.arguments);

      if (result.id === "sam") {
        return true;
      } else {
        return false;
      }
    }),
    // pipeline's request mapping template
    "{}",
    // function's request mapping template
    `${VTL.CircuitBreaker}
#set($v1 = {\"version\": \"2018-05-29\", \"operation\": \"Invoke\", \"payload\": $context.arguments})
$util.toJson($v1)`,
    // function's response mapping template
    `#set( $context.stash.result = $context.result )
{}`,
    // response mapping template
    `#if($context.stash.return__flag)
  #return($context.stash.return__val)
#end
#set($v1 = $context.stash.result.id == 'sam')
#if($v1)
#set($context.stash.return__val = true)
#set($context.stash.return__flag = true)
#return($context.stash.return__val)
#else
#set($context.stash.return__val = false)
#set($context.stash.return__flag = true)
#return($context.stash.return__val)
#end`
  ));

test("call function omitting optional arg", () =>
  appsyncTestCase(
    reflect((context: AppsyncContext<{ arg: string }>) => {
      return fn2(context.arguments);
    }),
    // pipeline's request mapping template
    "{}",
    // function's request mapping template
    `${VTL.CircuitBreaker}
#set($v1 = {\"version\": \"2018-05-29\", \"operation\": \"Invoke\", \"payload\": $context.arguments})
$util.toJson($v1)`,
    // function's response mapping template
    `#set( $context.stash.return__flag = true )
#set( $context.stash.return__val = $context.result )
{}`,
    // response mapping template
    `#if($context.stash.return__flag)
  #return($context.stash.return__val)
#end`
  ));

test("call function including optional arg", () =>
  appsyncTestCase(
    reflect((context: AppsyncContext<{ arg: string }>) => {
      return fn2({ arg: context.arguments.arg, optional: "hello" });
    }),
    // pipeline's request mapping template
    "{}",
    // function's request mapping template
    `${VTL.CircuitBreaker}
#set($v1 = {})
$util.qr($v1.put('arg', $context.arguments.arg))
$util.qr($v1.put('optional', 'hello'))
#set($v2 = {\"version\": \"2018-05-29\", \"operation\": \"Invoke\", \"payload\": $v1})
$util.toJson($v2)`,
    // function's response mapping template
    `#set( $context.stash.return__flag = true )
#set( $context.stash.return__val = $context.result )
{}`,
    // response mapping template
    `#if($context.stash.return__flag)
  #return($context.stash.return__val)
#end`
  ));

test("call function including with no parameters", () =>
  appsyncTestCase(
    reflect(() => {
      return fn3();
    }),
    // pipeline's request mapping template
    "{}",
    // function's request mapping template
    `${VTL.CircuitBreaker}
#set($v1 = {\"version\": \"2018-05-29\", \"operation\": \"Invoke\", \"payload\": $null})
$util.toJson($v1)`,
    // function's response mapping template
    `#set( $context.stash.return__flag = true )
#set( $context.stash.return__val = $context.result )
{}`,
    // response mapping template
    `#if($context.stash.return__flag)
  #return($context.stash.return__val)
#end`
  ));

test("call function including with void result", () =>
  appsyncTestCase(
    reflect((context: AppsyncContext<{ arg: string }>) => {
      return fn4(context.arguments);
    }),
    // pipeline's request mapping template
    "{}",
    // function's request mapping template
    `${VTL.CircuitBreaker}
#set($v1 = {\"version\": \"2018-05-29\", \"operation\": \"Invoke\", \"payload\": $context.arguments})
$util.toJson($v1)`,
    // function's response mapping template
    `#set( $context.stash.return__flag = true )
#set( $context.stash.return__val = $context.result )
{}`,
    // response mapping template
    `#if($context.stash.return__flag)
  #return($context.stash.return__val)
#end`
  ));

test("function inline arrow closure", () => {
  new Function(stack, "inline", async (p: string) => p);
});

test("function block closure", () => {
  new Function(stack, "block", async (p: string) => {
    return p;
  });
});
