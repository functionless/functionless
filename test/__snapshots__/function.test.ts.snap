// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`call function 1`] = `
[
  "{}",
  "#if($context.stash.return__flag)
  #return($context.stash.return__val)
#end
#set($v1 = {"version": "2018-05-29", "operation": "Invoke", "payload": $context.arguments})
$util.toJson($v1)",
  "#set( $context.stash.return__flag = true )
#set( $context.stash.return__val = $context.result )
{}",
  "#if($context.stash.return__flag)
  #return($context.stash.return__val)
#end",
]
`;

exports[`call function and conditional return 1`] = `
[
  "{}",
  "#if($context.stash.return__flag)
  #return($context.stash.return__val)
#end
#set($v1 = {"version": "2018-05-29", "operation": "Invoke", "payload": $context.arguments})
$util.toJson($v1)",
  "#set($context.stash.result = $context.result)
{}",
  "#if($context.stash.return__flag)
  #return($context.stash.return__val)
#end
#set($v1 = $context.stash.result.id == 'sam')
#if($v1)
#set($context.stash.return__val = true)
#set($context.stash.return__flag = true)
#return($context.stash.return__val)
#else
#set($context.stash.return__val = false)
#set($context.stash.return__flag = true)
#return($context.stash.return__val)
#end",
]
`;

exports[`call function including optional arg 1`] = `
[
  "{}",
  "#if($context.stash.return__flag)
  #return($context.stash.return__val)
#end
#set($v1 = {})
$util.qr($v1.put('arg', $context.arguments.arg))
$util.qr($v1.put('optional', 'hello'))
#set($v2 = {"version": "2018-05-29", "operation": "Invoke", "payload": $v1})
$util.toJson($v2)",
  "#set( $context.stash.return__flag = true )
#set( $context.stash.return__val = $context.result )
{}",
  "#if($context.stash.return__flag)
  #return($context.stash.return__val)
#end",
]
`;

exports[`call function including with no parameters 1`] = `
[
  "{}",
  "#if($context.stash.return__flag)
  #return($context.stash.return__val)
#end
#set($v1 = {"version": "2018-05-29", "operation": "Invoke", "payload": $null})
$util.toJson($v1)",
  "#set( $context.stash.return__flag = true )
#set( $context.stash.return__val = $context.result )
{}",
  "#if($context.stash.return__flag)
  #return($context.stash.return__val)
#end",
]
`;

exports[`call function including with void result 1`] = `
[
  "{}",
  "#if($context.stash.return__flag)
  #return($context.stash.return__val)
#end
#set($v1 = {"version": "2018-05-29", "operation": "Invoke", "payload": $context.arguments})
$util.toJson($v1)",
  "#set( $context.stash.return__flag = true )
#set( $context.stash.return__val = $context.result )
{}",
  "#if($context.stash.return__flag)
  #return($context.stash.return__val)
#end",
]
`;

exports[`call function omitting optional arg 1`] = `
[
  "{}",
  "#if($context.stash.return__flag)
  #return($context.stash.return__val)
#end
#set($v1 = {"version": "2018-05-29", "operation": "Invoke", "payload": $context.arguments})
$util.toJson($v1)",
  "#set( $context.stash.return__flag = true )
#set( $context.stash.return__val = $context.result )
{}",
  "#if($context.stash.return__flag)
  #return($context.stash.return__val)
#end",
]
`;
